From 27229dccf89edabd16f5a13e13a30b486384df7f Mon Sep 17 00:00:00 2001
From: Alexei Podtelezhnikov <apodtele@gmail.com>
Date: Sun, 31 May 2026 08:07:38 -0400
Subject: [PATCH] * src/autofit/afgsub.c (af_validate_coverage): Reject
 overlaps.

---
 src/autofit/afgsub.c | 13 ++++++-------
 1 file changed, 6 insertions(+), 7 deletions(-)

diff --git a/src/autofit/afgsub.c b/src/autofit/afgsub.c
index efcba18ae..dfdb99641 100644
--- a/src/autofit/afgsub.c
+++ b/src/autofit/afgsub.c
@@ -67,6 +67,8 @@
     {
       FT_UInt   rangeCount = FT_NEXT_USHORT( p );
       FT_Byte*  limit      = p + rangeCount * 6;
+      FT_Int    startGlyphID;
+      FT_Int    endGlyphID = -1;
 
 
       if ( table_limit < limit )
@@ -74,15 +76,12 @@
 
       while ( p < limit )
       {
-        FT_UInt  startGlyphID = FT_NEXT_USHORT( p );
-        FT_UInt  endGlyphID   = FT_NEXT_USHORT( p );
-
-
-        if ( startGlyphID > endGlyphID )
+        startGlyphID = FT_NEXT_USHORT( p );
+        if ( startGlyphID <= endGlyphID )
           return FALSE;
 
-        /* If `endGlyphID` is 0xFFFF it must be the last element. */
-        if ( endGlyphID == 0xFFFF && p < limit )
+        endGlyphID = FT_NEXT_USHORT( p );
+        if ( endGlyphID < startGlyphID )
           return FALSE;
 
         count += endGlyphID - startGlyphID + 1;
